pkg/resource/api.go

```diff
@@ -47,90 +47,118 @@ func NewAPIPatchingApplicator(c client.Client) *APIPatchingApplicator {
 	return &APIPatchingApplicator{client: c}
 }
 
+// WithLogger sets the logger on the APIPatchingApplicator.
+func (a *APIPatchingApplicator) WithLogger(l logging.Logger) *APIPatchingApplicator {
+	a.optionalLog = l
+	return a
+}
```

Comment:
This is introducing a second way to pass options (builder style) in addition to the variadic `ApplyOption` style supported by the `Apply` method.

Given that:
* As you mention `LogDiff` is expensive, so we want it to be done optionally.
* `LogDiff` is called right after where we loop over the variadic option functions.
* `LogDiff` has a very similar signature to the variadic option functions.

Could `LogDiff` itself be an option - e.g. `p.Apply(ctx, o, WithLogDiff(l logging.Logger))`?

Reply:
I had that in the beginning and forgot why I reverted it.